Broken Metrics detector

paologallinaharbur opened the pull request:
## wrong-type-detector

The tool aims to detect from a prometheus output the wrong metrics and provides the rules in order to fix them.
To detect the broken metrics, i.e. not following
```
metricName_bucket is stored as a New Relic count metric type.
metricName_count is stored as a New Relic count metric type.
metricName_total is stored as a New Relic count metric type.
```

we run a query similar to:
```
 FROM  Metric SELECT uniqueCount(metricName) where 
 (nrMetricType='gauge' and (metricName LIKE '%_total' or metricName LIKE '%_count'   )) or 
 (nrMetricType='count' and (metricName NOT LIKE '%_total' and metricName NOT LIKE '%_count'   ))  
 and collector.name = 'nri-prometheus' limit max
```


Currently, the tool focus on counter only since the summary one should not break since they are autogenerated.
However, we can add that if we notice that it is useful:
> metricName_sum is stored as a New Relic summary metric type.
### Usage


```
go run ./cmd/wrong-type-detector/main.go  -h              
Usage of main:
  -outputMetrics string
        output for the broken metrics (default "output/broken.yaml")
  -outputRules string
        output for the rules to fix broken metrics (default "output/rules.yaml")
  -promInput string
        source for the prometheus metrics (default "./powerdns.prom")
```

Simply run:
```
$ go run ./cmd/wrong-type-detector/main.go 
2022/10/12 16:53:46 broken metrics detected in "./powerdns.prom", placed in "output/broken.yaml" and the rules fixen them in "output/rules.yaml"
```

marcsanmi commented on tools/cmd/wrong-type-detector/main.go:
```diff
@@ -0,0 +1,145 @@
+
+		_, err = fmt.Fprintf(file, genericRule, m.Name, nrMetricType)
+		if err != nil {
+			if err != nil {
```
Spare one `err != nil`?

kang-makes commented on tools/cmd/wrong-type-detector/main.go:
```diff
@@ -0,0 +1,150 @@
+package main
```
Can we change the path to this file?

I would change it to `tools/wrong-type-detector/cmd/main.go` as it is the `cmd` of the tool `wrong-type-detector`. I would even remove the `cmd` part...

paologallinaharbur commented:
I rebased the PR and fixed some of the issues spotted during the review. There are still no test and it is still a big file since it is a "temporal" internal tool

kang-makes reviewed (approved):
I hope it's as temporary as you claim.

Just a NIT, I would remove all the examples and keep a couple, one that has all the metrics ok and one completely broken.

On the other hand, LGTM. Thanks for the tool :)

paologallinaharbur commented:
I'll remove a couple that @alvarocabanas added 😈
